extension/goext/types.go

```diff
@@ -0,0 +1,298 @@
+// Copyright (C) 2017 NTT Innovation Institute, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package goext
+
+import "encoding/json"
+
+type MaybeState int
+
+const (
+	MaybeUndefined MaybeState = iota
+	MaybeNull
+	MaybeNotNull
+)
+
+// Maybe is a 3-value state keeper
+//
+// Add it to your custom structure as a anonymous field to make structure 3-value
+// 3-value structures are marshalled into JSON value based on their state
+// Structures in:
+// - Undefined state won't appear in JSON at all
+// - Null state will appear in JSON as null
+// - NotNull state will appear in JSON as they would normally be marshalled
+//
+// Example:
+// Given structures:
+// type TestSuite struct {
+// 	 goext.Maybe
+//	 ID string `json:"id"`
+// }
+//
+// type Test struct {
+// 	 ID string `json:"id"`
+//	 Name string  `json:"name"`
+//	 TestSuite TestSuite `json:"test_suite"`
+// }
+//
+// json.Marshal(env.Util().ResourceToMap(&Test{ID: "some-id", Name:"My resource"}))
+// Will produce {"id": "some-id", "name": "My resource"}
+//
+// json.Marshal(env.Util().ResourceToMap(&Test{ID: "some-id", Name:"My resource"}, TestSuite: MakeNullTestSuite()))
+// Will produce {"id": "some-id", "name": "My resource", "test_suite": null}
+//
+// json.Marshal(env.Util().ResourceToMap(&Test{ID: "some-id", Name:"My resource"}, TestSuite: MakeTestSuite("test-suite-id")))
+// Will produce {"id": "some-id", "name": "My resource", "test_suite": {"id": "test-suite-id"}}
+//
+type Maybe struct {
+	MaybeState MaybeState
+}
+
+// IsUndefined returns whether value is undefined
+func (m Maybe) IsUndefined() bool {
+	return m.MaybeState == MaybeUndefined
+}
+
+// IsNull returns whether value is null
+func (m Maybe) IsNull() bool {
+	return m.MaybeState == MaybeNull
+}
+
+// IsNotNull returns whether value is defined and not null
+func (m Maybe) IsNotNull() bool {
+	return m.MaybeState == MaybeNotNull
+}
+
+// MaybeString represents 3-valued string
+type MaybeString struct {
```

Comment:
the API became hard to use correctly and easy to use incorrectly. consider:
```
str := MakeNullString()
...
// later somewhere in the code
str.Value = "some value"
// str.MaybeState is still MaybeNull. I'd expect MaybeNotNull, because a value has been assigned
```
I suggest adding Get/SetValue methods and making Value private. SetValue should set state to MaybeNotNull, GetValue should return the value + state or value + error

Reply:
+1

Reply:
I see, that is a big problem for custom structures embedding goext.Maybe. We can't enforce user to make all fields as private, and rely that all setters will change state of object. Even if this is possible, such object would be very painful in use. I'll revert last change (embedding Maybe in custom structs), and make primitive types immutable.

Reply:
ok, got it